### ISSUE TITLE
Metadata of all hashed metagenomes?

### ISSUE THREAD

**Issue by CalvinRusley** [sourmash-bio/branchwater#32]
First of all, thank you for designing such a useful tool!
Is there any plan to make the list of all hashed SRA accessions with associated metadata (i.e. lat/long) available?

**Comment by VDaric** [comment]
I have the same question.

A particular project doesn’t appear in my query results. I’m wondering if it’s because my species of interest isn’t included there, or because the project isn’t in the database.

Thanks !

**Comment by luizirber** [comment]
Hey @CalvinRusley and @VDaric , sorry for the delay in answering this...

We don't have an API endpoint to return all the associated metadata yet, but all the metadata for the current index is available as a parquet file in https://farm.cse.ucdavis.edu/~irber/branchwater/20241128-metadata.parquet

There are also three API endpoints (currently only deployed in the dev server, which is `scaled=100,000` instead of `scaled=1,000`, but has the same metadata) for returning:

- stats: mostly useful for `n_datasets`, the number of datasets in the current index: 
```
$ curl https://api.branchwater-dev.sourmash.bio/metadata/stats
{"ksize":21,"scaled":100000,"threshold":0,"n_datasets":1161119}
```

- accessions: the SRA run IDs currently indexed (beware, this is 1M+ lines long 😅 ):
```
$ curl https://api.branchwater-dev.sourmash.bio/metadata/accessions | head
ERR4170321
SRR20082975
SRR15343452
SRR20006711
SRR17780139
SRR8303175
ERR6997859
ERR2046471
SRR21705884
SRR14371704
```

- manifest: the [sourmash manifest](https://sourmash.readthedocs.io/en/latest/sourmash-internals.html#manifests) used for building the index (also pretty big)
```
$ curl https://api.branchwater-dev.sourmash.bio/metadata/manifest | head
internal_location,md5,md5short,ksize,moltype,num,scaled,n_hashes,with_abundance,name,filename
sigs/ERR4170321.sig,83a5c54d156aa0c5bfbc3f4c61ca1704,83a5c54d,21,DNA,0,100000,2960,1,ERR4170321,-
sigs/SRR20082975.sig,40a473c95e801d9b32b8bfc30c43491c,40a473c9,21,DNA,0,100000,199745,1,SRR20082975,-
sigs/SRR15343452.sig,002b96648037c6e2a56de6d804087eb3,002b9664,21,DNA,0,100000,367414,1,SRR15343452,-
sigs/SRR20006711.sig,9fedfb6cee6a2e0a8e6ef1ffa9a538a1,9fedfb6c,21,DNA,0,100000,37325,1,SRR20006711,-
sigs/SRR17780139.sig,480e92a4f55c910235072f792b8ee418,480e92a4,21,DNA,0,100000,58261,1,SRR17780139,-
sigs/SRR8303175.sig,da76545da06ed4b87547b10bd239a5d0,da76545d,21,DNA,0,100000,97945,1,SRR8303175,-
sigs/ERR6997859.sig,faa612cf7ddc6fb740bc204db5f39246,faa612cf,21,DNA,0,100000,15963,1,ERR6997859,-
sigs/ERR2046471.sig,8f8f6f93a52047504eb69105bb963e39,8f8f6f93,21,DNA,0,100000,621,1,ERR2046471,-
sigs/SRR21705884.sig,dc92973452c87ccbdea96404863a502d,dc929734,21,DNA,0,100000,1866195,1,SRR21705884,-
```

---

Probably need a 4th one, for the full metadata as a parquet file. But for now there is the download link I posted above.

Hope this helps!

**Comment by VDaric** [comment]
Thank you very much for your answer.

I think I might be missing something.

I submitted a cyanobacterial genome as the query. In the CSV result file I received, one of the accessions listed was ['SRR16348820](https://www.ebi.ac.uk/ena/browser/view/SRR16348820)'. The "_Library Strategy_" field in its metadata has the value **RNA-Seq**. I’m actually interested in meta-transcriptomes as well.

I downloaded the metadata file from [https://api.branchwater-dev.sourmash.bio/metadata/manifest](https://api.branchwater-dev.sourmash.bio/metadata/manifest).

```
$ grep SRR16348820 sourmash.manifest
sigsSRR16348820.sig,e94cb48934de2d81bd4f54cc8b15cf47,e94cb489,21,DNA,0,100000,2169475,1,SRR16348820,-
```


The value in the "_moltype_" column is **DNA**, whereas I would have expected **RNA**.

More broadly, I’m trying to determine the proportion of RNA versus DNA data available in the database. Understanding this distribution would help me assess whether I can use [https://branchwater.jgi.doe.gov/](https://branchwater.jgi.doe.gov/) to explore meta-transcriptomic data.

Thanks !

**Comment by luizirber** [comment]
> I submitted a cyanobacterial genome as the query. In the CSV result file I received, one of the accessions listed was ['SRR16348820](https://www.ebi.ac.uk/ena/browser/view/SRR16348820)'. The "_Library Strategy_" field in its metadata has the value **RNA-Seq**. I’m actually interested in meta-transcriptomes as well.
> 
> I downloaded the metadata file from https://api.branchwater-dev.sourmash.bio/metadata/manifest.
> 
> ```
> $ grep SRR16348820 sourmash.manifest
> sigsSRR16348820.sig,e94cb48934de2d81bd4f54cc8b15cf47,e94cb489,21,DNA,0,100000,2169475,1,SRR16348820,-
> ```
> 
> The value in the "_moltype_" column is **DNA**, whereas I would have expected **RNA**.

Not quite... `DNA` means calculated from nucleotides (DNA or RNA), while protein/dayhoff/hp assume aminoacids. Or you can translate DNA to aminoacid with `sourmash sketch translate`.

More info about moltypes: https://sourmash.readthedocs.io/en/latest/sourmash-internals.html#molecule-types-dna-protein-dayhoff-and-hydrophobic-polar

And for a gut check, I took [the SRA run you linked (SRR16348820)](https://trace.ncbi.nlm.nih.gov/Traces/?view=run_browser&page_size=10&acc=SRR16348820&display=reads) and checked the reads to make sure they are `ACTG` (and not `ACUG`), because we have [a pretty old PR](https://github.com/sourmash-bio/sourmash/pull/148) mentioned [frequently](https://github.com/COMBINE-lab/cuttlefish/issues/58) in discussions on how to deal with ambiguous bases (N) and how to deal with non-standard alphabets (discard k-mer, unless `force=True` is set). So DNA or RNA behave the same in this case.

> More broadly, I’m trying to determine the proportion of RNA versus DNA data available in the database. Understanding this distribution would help me assess whether I can use https://branchwater.jgi.doe.gov/ to explore meta-transcriptomic data.

Nowadays most of the metatranscriptomic datasets in branchwater are off-target from the SRA metadata selection process, [since we don't select them directly](https://github.com/sourmash-bio/branchwater/issues/24#issuecomment-2067814713). But we have the signatures in wort, since they are built daily from [this query](https://github.com/sourmash-bio/wort/blob/6b0bf44d633827dfe1f67e48a81dbb6de357874a/machine/wort-web/daily-sra-biopython.py#L16-L18) to the SRA:

```python
query_dna = f'("{day_to_check}"[PDAT] : "{end_date}"[PDAT]) AND "biomol dna"[Properties] NOT amplicon[All Fields] NOT "Metazoa"[orgn:__txid33208] NOT "Streptophyta"[orgn:__txid35493] NOT "METAGENOMIC"[Source]'
query_metagenomic = f'("{day_to_check}"[PDAT] : "{end_date}"[PDAT]) NOT amplicon[All Fields] AND "METAGENOMIC"[Source]'
query_metatranscriptomic = f'("{day_to_check}"[PDAT] : "{end_date}"[PDAT]) NOT amplicon[All Fields] AND "METATRANSCRIPTOMIC"[Source]'
```

The fun starts when they change status, sometimes metadata is updated (but not the sequences), and so it might show up as metagenomic one day, and metatranscriptomic later. See [the second half of this comment](https://github.com/sourmash-bio/branchwater/issues/24#issuecomment-2068194154) for more details.

---

I [extended the query](https://www.ncbi.nlm.nih.gov/sra/?term=(%221950%2F01%2F01%22%5BPDAT%5D+%3A+%222024%2F04%2F20%22%5BPDAT%5D)+NOT+amplicon%5BAll+Fields%5D+AND+(%22METAGENOMIC%22%5BSource%5D+OR+%22METATRANSCRIPTOMIC%22%5BSource%5D)) to include both metagenomic and metatranscriptomic sources, and went from
```
    Source
        DNA(1,148,562)
```
to
```
    Source
        DNA(1,148,562)
        RNA(98,362)
```
so metatranscriptomic is about 8% of metagenomic sources. I should probably include them in the next index update to make your analysis easier...

**Comment by VDaric** [comment]
That would be amazing, thanks so much! Really appreciate your help with this.

**Comment by CalvinRusley** [comment]
@luizirber I see that the name of that file is 20241128 – is this still current? I've found at least 64 SRA accessions in my results from last week that aren't in the metadata parquet file.

It's also worth noting that some (i.e. SRR24717120), while they don't show up when searched for in the SRA, do apparently exist under SRX prefixes, but in the case of SRX24717120 is not public.
